src/esp/assets/ResourceManager.cpp

```diff
@@ -1591,8 +1616,76 @@ int ResourceManager::loadNavMeshVisualization(esp::nav::PathFinder& pathFinder,
   return navMeshPrimitiveID;
 }  // ResourceManager::loadNavMeshVisualization
 
+namespace {
+
+/**
+ * @brief Make sure we do not add an attribute that already exists in source
+ * material - this will assert.
+ * @tparam the type of the value to be added
+ * @param material Source material for new attributes array
+ * @param newAttributes The array of attributes to be used to build a new
+ * Mn::Trade::MaterialData.
+ * @param matAttr The Mn::Trade::MaterialAttribute tag describing the attribute
+ * to add
+ * @param value A reference to the value to add.
+ */
+template <typename T>
+void safeAppend(
+    const Cr::Containers::Optional<Mn::Trade::MaterialData>& material,
+    Cr::Containers::Array<Mn::Trade::MaterialAttributeData>& newAttributes,
+    const MnMatAttr& matAttr,
+    T& value) {
+  if (!material->hasAttribute(matAttr)) {
+    arrayAppend(newAttributes, {matAttr, std::move(value)});
+  }
+}
+
+/**
+ * @brief given passed @ref metadata::attributes::ObjectInstanceShaderType @p
+ * typeToCheck and given @ref Magnum::Trade::MaterialData, verify that the
+ * material's intrinsic type is the same as inferred by @p typeToCheck. Ignore
+ * flat, since all shaders already support flat.
+ * @param typeToCheck The type of shader being specified.
+ * @param materialData The imported material to check for type
+ * @return Whether the imported material's supported types include one
+ * congruient with the specified shader type.
+ */
+bool compareShaderTypeToMnMatType(
+    const ShaderTypeEnum typeToCheck,
+    const Cr::Containers::Optional<Mn::Trade::MaterialData>& materialData) {
+  switch (typeToCheck) {
+    case ShaderTypeEnum::Phong: {
+      bool compRes =
+          ((materialData->types() & Mn::Trade::MaterialType::Phong) ==
+           Mn::Trade::MaterialType::Phong);
```

Comment:
FYI the EnumSet class has a [`operator>=`](https://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1EnumSet.html#aeb270daa9a8afb3a80e57ef2cb2f8314) which resembles a ⊇ set operation in mathematics, which would shorten the above to

```suggestion
          materialData->types() >= Mn::Trade::MaterialType::Phong);
```

It might be however confusing to some people because this does a different thing than it would do with pure integers. The `MaterialType` enum is just single bits so even

```suggestion
          !!(materialData->types() & Mn::Trade::MaterialType::Phong);
```

would work for it right now (the !! for forcing an explicit bool conversion), alternatively

```suggestion
          bool(materialData->types() & Mn::Trade::MaterialType::Phong);
```

if you don't like the excessive yelling :)

Same below.

Reply:
wow. the enum op overload is really cool, ty.